docs/rfcs/025-generation-numbers.md

```diff
@@ -0,0 +1,468 @@
+   logical node 0, it's just a different physical node. So we have an O(tenant_count) communication
+   to the pageserver, but not an O(tenant_count) write to the database.
+1. S3 writes continue from nodes 0a and 0b, but the writes do not collide due to different
+   node generation in the suffix, and the writes from node 0a are not visible to the rest
+   of the system because endpoints are reading only from node 0b.
+
+# Appendix B: interoperability with other features
+
+## Sharded Keyspace
+
+The design in this RFC maps neatly to a sharded keyspace design where subsets of the key space
+for a tenant are assigned to different pageservers:
+
+- the "unit of work" for attachments becomes something like a TenantShard rather than a Tenant
+- TenantShards get generation numbers just as Tenants do.
+- Write workload (ingest, compaction) for a tenant is spread out across pageservers via
+  TenantShards, but each TenantShard still has exactly one valid writer at a time.
+
+## Read replicas
+
+For historical reads to LSNs below the remote persistent LSN, any node may act as a reader at any
+time: remote data is logically immutable data, and the use of deferred deletion in this RFC helps
+mitigate the fact that remote data is not _physically_ immutable (i.e. the actual data for a given
+page moves around as compaction happens).
+
+A read replica needs to be aware of generations in remote data in order to read the latest
+metadata (find the index_part.json with the latest suffix). This does not have to involve any
+control plane interaction: a read replica node can scan remote index files to find the latest one. However,
+it may make sense to send notifications from the control plane to avoid the need for read replicas
+to poll S3 to detect when the writer for a particular tenant changed, or to expose this information
+via some intermediate pub/sub mechanism for async consumption by read replicas.
+
+## Seamless migration
+
+To make tenant migration totally seamless, we will probably want to intentionally double-attach
+a tenant briefly, serving reads from the old node while waiting for the new node to be ready.
+
+This doesn't conflict with this RFC. A node may go through a transfer routine where it flushes
+data to S3 and then goes into a readonly state pending the new node's active state. The attachment
+generation number would increment when the control plane decides to "commit" the transfer by making
+the new location the leader.
+
+This might be implemented as a specialization of "warm secondary locations" (below), where doing
+a clean migration is just creating a temporary warm secondary location.
+
+## Warm secondary locations
+
+To enable faster tenant movement after a pageserver is lost, we will probably want to spend some
+disk capacity on keeping standby locations populated with local disk data.
+
+There's no conflict between this RFC and that: implementing warm secondary locations on a per-tenant basis
```

Comment:
Note that in order for secondary to be warm it needs to know primary's working set. Otherwise failover process wont be significantly faster compared to attaching to a fresh instance

Reply:
Right.  I'm planning on writing a separate RFC to describe how warm secondary locations should work.